### PR TITLE
Add ESM contact prediction

### PR DESCRIPTION
**Pull request by Rocketknight1**
This PR adds the `ContactPredictionHead` for ESM (both PT and TF). I also need to update some weights on our uploaded models to support this!

### CODE REVIEW

File: src/transformers/models/esm/modeling_esm.py [hunk 1]
```diff
@@ -111,6 +128,41 @@ def forward(self, q: torch.Tensor, k: torch.Tensor) -> Tuple[torch.Tensor, torch
...
+        eos_mask = tokens.ne(self.eos_idx).to(attentions)
+        eos_mask = eos_mask.unsqueeze(1) * eos_mask.unsqueeze(2)
+        attentions = attentions * eos_mask[:, None, None, :, :]
+        attentions = attentions[..., :-1, :-1]
+        # remove cls token attentions
+        attentions = attentions[..., 1:, 1:]
+        batch_size, layers, heads, seqlen, _ = attentions.size()
+        attentions = attentions.view(batch_size, layers * heads, seqlen, seqlen)
+
+        # features: B x C x T x T
```
Comment (sgugger) [inline src/transformers/models/esm/modeling_esm.py:157]: As always, would love to have full names and not one-letters to understand what's going on ;-)

File: src/transformers/models/esm/modeling_tf_esm.py [hunk 2]
```diff
@@ -71,6 +71,23 @@ def apply_rotary_pos_emb(x, cos, sin):
     return (x * cos) + (rotate_half(x) * sin)
 
 
+def symmetrize(x):
+    "Make layer symmetric in final two dimensions, used for contact prediction."
+    return x + tf.linalg.matrix_transpose(x)  # Transposes last two dimensions only
+
+
+def apc(x):
```
Comment (sgugger) [inline src/transformers/models/esm/modeling_tf_esm.py:79]: Same as in the PyTorch file.

File: src/transformers/models/esm/modeling_tf_esm.py [hunk 3]
```diff
@@ -115,6 +132,43 @@ def call(self, q: tf.Tensor, k: tf.Tensor) -> Tuple[tf.Tensor, tf.Tensor]:
...
+        eos_mask = tf.cast(tokens != self.eos_idx, attentions.dtype)
+        eos_mask = tf.expand_dims(eos_mask, 1) * tf.expand_dims(eos_mask, 2)
+        attentions = attentions * eos_mask[:, None, None, :, :]
+        attentions = attentions[..., :-1, :-1]
+        # remove cls token attentions
+        attentions = attentions[..., 1:, 1:]
+        batch_size, layers, heads, seqlen, _ = shape_list(attentions)
+        attentions = tf.reshape(attentions, (batch_size, layers * heads, seqlen, seqlen))
+
+        # features: B x C x T x T
```
Comment (sgugger) [inline src/transformers/models/esm/modeling_tf_esm.py:166]: Here too

**Review by sgugger — APPROVED** [review]
Thanks for adding this, just a few styling nits.

**Comment by Rocketknight1** [comment]
@sgugger All suggestions included!

**Pull request merged**